File: money_service.js
const { getPromptAsFloat, getPromptAsNumber } = require('./utils');

let balance = 0;
let latestActions = [];
let currency = "Ariary";

function print(message) {
    console.log(message);
}

function checkBalance() {
    print(`Your current balance is: ${balance} ${currency}`);
    latestActions.push(`You checked your balance. That was ${balance} ${currency}`);
}

function retrieveMoney() {
    print(`How much ${currency} do you want to get out of your wallet?`);
    const wantedMoney = getPromptAsFloat();
    if (balance > wantedMoney && wantedMoney >= 0) {
        balance = (balance - wantedMoney);
        const message = `You took ${wantedMoney} ${currency} out.`;
        print(message);
        latestActions.push(message);
    } else {
        print("You can't do that.");
        print(`Your current balance is: ${balance} ${currency}`);
    }
}

function storeMoney() {
    print(`How much ${currency} do you want to add?`);
    const moneyToAdd = getPromptAsFloat();
    if (moneyToAdd >= 0) {
        balance = (balance + moneyToAdd);
        const message = `You added ${moneyToAdd} ${currency} to your balance.`;
        print(message);
        latestActions.push(message);
    } else {
        print("You can't do that.");
    }
}

function convertCurrency() {
    print("Where are you now? Convert to your preferred currency.");
    print("1. To Euro.");
    print("2. To Dollar.");
    print("3. To Ariary.");
    const choice = getPromptAsNumber();
    switch (choice) {
        case 1:
            convertToChosenCurrency(currency, "Euro");
            break;
        case 2:
            convertToChosenCurrency(currency, "Dollar");
            break;
        case 3:
            convertToChosenCurrency(currency, "Ariary");
            break;
        default:
    
    }
}

function convertToChosenCurrency(currentCurrency, chosenCurrency) {
    if (currentCurrency === chosenCurrency) {
        print("It's already the current currency.");
    } else {
        currency = (chosenCurrency);
        const message = `You choose ${chosenCurrency} as currency.`;
        print(message);
        latestActions.push(message);
        if (balance > 0) {
            switch (currentCurrency) {
                case "Ariary":
                    chosenCurrency === "Dollar" ? balance = (balance * 0.00022) :
                        chosenCurrency === "Euro" ? balance = (balance * 0.00020) : null;
                    break;
                case "Dollar":
                    chosenCurrency === "Ariary" ? balance = (balance * 4582) :
                        chosenCurrency === "Euro" ? balance = (balance * 0.93) : null;
                    break;
                case "Euro":
                    chosenCurrency === "Ariary" ? balance = (balance * 4582) :
                        chosenCurrency === "Dollar" ? balance = (balance * 1.08) : null;
                    break;
            }
        }
    }
}

function history() {
    print("Your latest actions from recent to older:");
    latestActions.slice(-5).reverse().forEach(print);
}


module.exports = {
    checkBalance,
    retrieveMoney,
    storeMoney,
    convertCurrency,
    history,
    print
};

File: src/com/wallet/service/MoneyService.java
package com.wallet.service;

import java.util.Collections;
import java.util.List;

import com.wallet.Utils;
import com.wallet.abstraction.Compartment;

public class MoneyService implements Compartment {
    private double balance;
    private List<String> latestActions;

    private String currency;
    static final String cannotDo = "You can't do that.";

    final WalletService walletService = new WalletService();


    public MoneyService(double balance, List<String> latestActions, String currency) {
        this.balance = balance;
        this.latestActions = latestActions;
        this.currency = currency;
    }


    public String getCurrency() {
        return currency;
    }

    public void setCurrency(String currency) {
        this.currency = currency;
    }

    public void getBalance() {
        Utils.print("Your current balance is: "+ this.balance + " " + currency);
    }

    public void setLatestActions(String actionToAdd) {
        this.latestActions.add(actionToAdd);
    }

    public void setBalance(double balance) {
        this.balance = balance;
    }

    public void checkBalance() {
        Utils.print("Your current balance is: "+ this.balance + " " + currency);
        setLatestActions("You checked your balance. That was " + this.balance + " " + currency);
        menu();
    }

    public void retrieveMoney() {
        Utils.print("How much " + currency + " do you want to get out of your wallet ?");
        double wantedMoney = Utils.doubleScanner();
        if(this.balance > wantedMoney && wantedMoney >= 0) {
            setBalance(this.balance - wantedMoney);
            String message = "You took " + wantedMoney + " " + currency + " out.";
            Utils.print(message);
            setLatestActions(message);
        } else {
            Utils.print(cannotDo);
            getBalance();
        }
        menu();
    }

    public void storeMoney() {
        Utils.print("How much " + currency + " do you want to add ?");
        double moneyToAdd = Utils.doubleScanner();
        if(moneyToAdd >= 0) {
            setBalance(this.balance + moneyToAdd);
            String message = "You added " + moneyToAdd + " " + currency + " to your balance.";
            Utils.print(message);
            setLatestActions(message);
        } else {
            Utils.print(cannotDo);
        }
        menu();
    }


    public void convertCurrency(){
        Utils.print("Where are you now ? Convert to your preferred currency.");
        Utils.print(new StringBuilder()
                .append("1. To Euro.")
                .append("\n")
                .append("2. To Dollar.")
                .append("\n")
                .append("3. To Ariary.")
                .append("\n")
                .toString());
        switch (Utils.intScanner()) {
            case 1 -> convertToChosenCurrency(currency, "Euro");
            case 2 -> convertToChosenCurrency(currency, "Dollar");
            case 3 -> convertToChosenCurrency(currency, "Ariary");
            default -> menu();
        }
        menu();
    }

    public void convertToChosenCurrency(String currentCurrency, String chosenCurrency) {
        if(currentCurrency.equals(chosenCurrency)){
            Utils.print("It's already the current currency.");
        } else {
            setCurrency(chosenCurrency);
            String message = "You choose " + chosenCurrency + " as currency.";
            Utils.print(message);
            latestActions.add(message);
            if(balance > 0) {
                switch (currentCurrency) {
                    case "Ariary" -> {
                        switch (chosenCurrency) {
                            case "Dollar" -> setBalance(balance * 0.00022);
                            case "Euro" -> setBalance(balance * 0.00020);
                        }
                    }
                    case "Dollar" -> {
                        switch (chosenCurrency) {
                            case "Ariary" -> setBalance(balance * 4582);
                            case "Euro" -> setBalance(balance * 0.93);
                        }
                    }
                    case "Euro" -> {
                        switch (chosenCurrency) {
                            case "Ariary" -> setBalance(balance * 4582);
                            case "Dollar" -> setBalance(balance * 1.08);
                        }
                    }
                }
            }
        }
    }

    public void history() {
        Utils.print("Your latest actions from recent to older:");
        Collections.reverse(latestActions);
        if(latestActions.size() > 5){
            for (int i = 0; i < 5; i++) {
                Utils.print(latestActions.get(i));
            }
        } else {
            latestActions.forEach(Utils::print);
        }
        menu();
    }

    @Override
    public void menu() {
        Utils.print(new StringBuilder()
                .append("1. Check your current balance.")
                .append("\n")
                .append("2. Take out money.")
                .append("\n")
                .append("3. Store money.")
                .append("\n")
                .append("4. Convert to another currency.")
                .append("\n")
                .append("5. Check your 5 latest transactions.")
                .append("\n")
                .append("6. See other objects")
                .toString());
        switch (Utils.intScanner()) {
            case 1 -> checkBalance();
            case 2 -> retrieveMoney();
            case 3 -> storeMoney();
            case 4 -> convertCurrency();
            case 5 -> history();
            case 6 -> precedentMenu();
            default -> menu();
        }
    }

    @Override
    public void precedentMenu() {
        walletService.menu();
    }
}
